File: hw1-search-ysabellaatehortua/Python/map.py
"""
Map class defines a graph representation for the city map.

Adam Eck
10/18/2021
"""

import mapnode


class Map():
    """
    Creates a new Map
    """
    def __init__(self):
        # create the fields
        self.startNode = None

        # create the collections
        self.nodes = []
        self.nodeMap = {}
        self.goalNodes = []


    """
    Adds a new MapNode to the Map

    node: the MapNode to add
    """
    def addMapNode(self, node):
        # add the state to the Map
        self.nodes.append(node)
        self.nodeMap[(node.lat, node.long)] = node


    """
    Writes the Map to a file

    filename: the name of the file in which to write the Map
    """
    def writeToFile(self, filename):
        file = open(filename, "w")

        # write the nodes to file
        file.write("Nodes:\n")
        for i in range(len(self.nodes)):
            node = self.nodes[i]
            line = "{}: {}\n".format(i, (node.lat, node.long))
            file.write(line)

        # write the edges to file
        file.write("\nEdges:\n")
        for i in range(len(self.nodes)):
            node = self.nodes[i]
            line = "{}: ".format(i)
            for j in range(len(node.neighbors)):
                neighbor = node.neighbors[j]
                if j > 0:
                    line += ", "
                line += "({}; {})".format(self.nodes.index(neighbor), round(node.costs[neighbor], 4))
            file.write("{}\n".format(line))

        # write the start node to file
        file.write("\nStart Node:\n")
        file.write("{}\n".format(self.nodes.index(self.startNode)))

        # finally, write the goal nodes to file
        file.write("\nGoal Nodes:\n")
        line = ""
        for i in range(len(self.goalNodes)):
            node = self.goalNodes[i]
            if (i > 0):
                line += ", "
            line += str(self.nodes.index(node))
        file.write("{}\n".format(line))

        # close the file
        file.close()

"""
Reads a Map from file

filename: the name of the file containing the Map
"""
def readFromFile(filename):
    # create the map to return
    map = Map()

    # read from the file
    file = open(filename, "r")
    lines = file.readlines()
    file.close()

    # process the lines
    inNodes = False
    inEdges = False
    inStart = False
    inGoals = False
    for line in lines:
        # does this line need processing?
        if (line == "\n"):
            continue
        elif (line.startswith("Nodes")):
            inNodes = True
            continue
        elif (line.startswith("Edges")):
            inNodes = False
            inEdges = True
            continue
        elif (line.startswith("Start")):
            inEdges = False
            inStart = True
            continue
        elif (line.startswith("Goal")):
            inStart = False
            inGoals = True
            continue

        # drop the newline
        line = line.replace("\n", "")

        # process the info on the line
        if (inNodes):
            split = line.split(":")
            loc = split[1]
            splitLoc = loc.replace("(", "").replace(")", "").replace(" ", "").split(",")

            nodeNum = int(split[0])
            lat = float(splitLoc[0])
            long = float(splitLoc[1])

            node = mapnode.MapNode(lat, long)
            map.addMapNode(node)
        elif (inEdges):
            split = line.split(":")
            neighbors = split[1]
            neighborsSplit = neighbors.replace(" ", "").split(",")

            nodeNum = int(split[0])
            node = map.nodes[nodeNum]
            for str in neighborsSplit:
                tupleSplit = str.replace("(", "").replace(")", "").replace(" ", "").split(";")

                neighborNum = int(tupleSplit[0])
                cost = float(tupleSplit[1])
                node.addNeighbor(map.nodes[neighborNum], cost)
        elif (inStart):
            startNum = int(line)
            map.startNode = map.nodes[startNum]
        elif (inGoals):
            line = line.replace(" ", "")
            split = line.split(",")
            for str in split:
                nodeNum = int(str)
                map.goalNodes.append(map.nodes[nodeNum])

    # return the read in map
    return map

File: hw1-search-ysabellaatehortua/Java/Map.java
import java.util.*;
import java.io.FileReader;
import java.io.BufferedReader;

/**
 * The {@link Map} class defines a graph representation for the city map.
 *
 * @author Adam Eck
 * @date 10/18/2021
 */
public class Map {
    /** The {@link MapNode} that the search should start from. */
    public MapNode startNode;

    /** The list of all {@link MapNode}s in the map. */
    public final List<MapNode> nodes;

    /** A reverse lookup for finding {@link MapNode}s by lat/lon. */
    public final HashMap<Double, HashMap<Double, MapNode>> nodeMap;

    /** The list of all {@link MapNode} goals. */
    public final List<MapNode> goalNodes;

    /**
     * Constructs a new {@link Map}.
     */
    public Map() {
        // create the fields
        startNode = null;

        // create the collections
        nodes = new ArrayList<>();
        nodeMap = new HashMap<>();
        goalNodes = new ArrayList<>();
    }

    /**
     * Adds a {@link MapNode} to the {@link Map}.
     *
     * @param node The {@link MapNode} to add
     */
    public void addMapNode(MapNode node) {
        if (!nodes.contains(node)) {
            nodes.add(node);

            if (!nodeMap.containsKey(node.lat)) {
                nodeMap.put(node.lat, new HashMap<>());
            }
            nodeMap.get(node.lat).put(node.lon, node);
        }
    }

    /**
     * A utility method for reading a {@link Map} from a given file.
     *
     * @param filename The path to the file to read the {@link Map} from
     *
     * @return The {@link Map} contained in {@code filename}
     */
    public static Map readFromFile(String filename) {
        Map map = new Map();

        // read from the file
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));

            boolean inNodes = false;
            boolean inEdges = false;
            boolean inStart = false;
            boolean inGoals = false;

            String line, loc, neighbors;
            String[] split, splitLoc, neighborsSplit, tupleSplit;
            int id, nodeNum, neighborNum;
            double lat, lon, cost;
            MapNode node;
            while (reader.ready()) {
                line = reader.readLine();

                // did we switch sections
                if ("".equals(line)) {
                    continue;
                } else if (line.startsWith("Nodes")) {
                    inNodes = true;
                    continue;
                } else if (line.startsWith("Edges")) {
                    inNodes = false;
                    inEdges = true;
                    continue;
                } else if (line.startsWith("Start")) {
                    inEdges = false;
                    inStart = true;
                    continue;
                } else if (line.startsWith("Goal")) {
                    inStart = false;
                    inGoals = true;
                    continue;
                }

                // process the info on the line
                if (inNodes) {
                    // split up the text
                    split = line.split(":");
                    loc = split[1];
                    splitLoc = loc.replaceAll("\\(", "").replaceAll("\\)", "")
                            .replaceAll(" ", "").split(",");

                    // parse the values
                    lat = Double.parseDouble(splitLoc[0]);
                    lon = Double.parseDouble(splitLoc[1]);

                    // create the node
                    node = new MapNode(lat, lon);
                    map.addMapNode(node);
                } else if (inEdges) {
                    split = line.split(":");
                    neighbors = split[1];
                    neighborsSplit = neighbors.replaceAll(" ", "").split(",");

                    nodeNum = Integer.parseInt(split[0]);
                    node = map.nodes.get(nodeNum);
                    for (String str : neighborsSplit) {
                        tupleSplit = str.replaceAll("\\(", "").replaceAll("\\)", "")
                                .replaceAll(" ", "").split(";");

                        neighborNum = Integer.parseInt(tupleSplit[0]);
                        cost = Double.parseDouble(tupleSplit[1]);
                        node.addNeighbor(map.nodes.get(neighborNum), cost);
                    }
                } else if (inStart) {
                    nodeNum = Integer.parseInt(line);
                    map.startNode = map.nodes.get(nodeNum);
                } else if (inGoals) {
                    split = line.replaceAll(" ", "").split(",");
                    for (String str : split) {
                        nodeNum = Integer.parseInt(str);
                        map.goalNodes.add(map.nodes.get(nodeNum));
                    }
                }
            }
        } catch (Exception ex) {
            System.err.println("Problem reading from file!");
            ex.printStackTrace();
            System.exit(-1);
        }

        return map;
    }
}
